File: warp/debug.py
"""
Debug-only blueprint — only registered when app.debug is True (DevelopmentSettings).
Provides endpoints to control "virtual time" for e2e tests.
"""
import flask
from . import utils

bp = flask.Blueprint('debug', __name__)


@bp.route('/debug/time', methods=['GET'])
def get_time():
    # Optional ?tz=<IANA> resolves now/today in that zone (PLAN per_plan_timezone
    # §7); without it, returns the clock in UTC (the feed/cache clock).
    tz = flask.request.args.get('tz') or 'UTC'
    if not utils.is_valid_iana(tz):
        return {"msg": "Invalid timezone"}, 400
    return {
        'now': utils.now(tz=tz),
        'today': utils.today(tz=tz),
        'offset_seconds': utils._debug_time_offset,
    }


@bp.route('/debug/set_time_offset', methods=['POST'])
def set_time_offset():
    """Shift utils.now() / utils.today() by offset_seconds.  Pass 0 to reset."""
    data = flask.request.get_json(silent=True) or {}
    offset = int(data.get('offset_seconds', 0))
    utils._debug_time_offset = offset
    return {
        'msg': 'ok',
        'offset_seconds': offset,
        'now': utils.now(tz='UTC'),
        'today': utils.today(tz='UTC'),
    }


@bp.route('/debug/set_language', methods=['POST'])
def set_language():
    """Switch the deployment fallback language for e2e (debug only).

    Sets DEFAULT_LANGUAGE at runtime (the per-user feature resolves a NULL
    user pref to this), and clears calendar_cache: cached feed rows are
    language-specific, so a switch must force regeneration. Accepts a
    `language` code ('de','en',...) or, for back-compat, a `language_file`
    path ('i18n/de.json') whose stem is used as the code. Reset with 'en'.
    """
    data = flask.request.get_json(silent=True) or {}
    code = data.get('language')
    if not code:
        lang_file = data.get('language_file') or 'i18n/en.json'
        code = lang_file.split('/')[-1].removesuffix('.json')
    # Reject a bogus code: the context processor / i18nUrl would build
    # i18n/<code>.json, which 404s and breaks the login page's i18n load.
    if code not in flask.current_app.config['LANGUAGES']:
        return {"msg": "unknown language", "code": 13}, 400
    flask.current_app.config['DEFAULT_LANGUAGE'] = code
    from warp.db import CalendarCache
    CalendarCache.delete().execute()
    return {
        'msg': 'ok',
        'language': code,
    }


@bp.route('/debug/endpoints', methods=['GET'])
def list_endpoints():
    """Return all non-debug registered routes — used by the e2e coverage test."""
    rules = [
        {
            'endpoint': r.endpoint,
            'rule': r.rule,
        }
        for r in flask.current_app.url_map.iter_rules()
        if r.endpoint != 'static' and not r.endpoint.startswith('debug.')
    ]
    return flask.jsonify(sorted(rules, key=lambda x: x['rule']))

File: e2e/helpers/debug.ts
import type { Page } from '@playwright/test';

/** Shift utils.now() / utils.today() on the server by offsetSeconds.
 *  Requires the Flask debug blueprint (only active in DevelopmentSettings).
 *
 *  Beware: shifting the clock forward by a day or more expires every login
 *  session (SESSION_LIFETIME) — log in again after calling this. The pristineDb
 *  fixture resets the offset before each test, so no manual cleanup is needed. */
export async function setTimeOffset(page: Page, offsetSeconds: number): Promise<void> {
  const resp = await page.request.post('/debug/set_time_offset', {
    data: { offset_seconds: offsetSeconds },
    headers: { 'Content-Type': 'application/json' },
  });
  if (!resp.ok()) {
    throw new Error(`set_time_offset failed: HTTP ${resp.status()}`);
  }
}

/** Reset server time to real wall-clock time. */
export async function resetTimeOffset(page: Page): Promise<void> {
  await setTimeOffset(page, 0);
}

/** Return { now, today, offset_seconds } from the server's perspective. */
export async function getServerTime(page: Page): Promise<{ now: number; today: number; offset_seconds: number }> {
  const resp = await page.request.get('/debug/time');
  return resp.json();
}

/** Convenience: advance by N whole days. */
export async function advanceDays(page: Page, days: number): Promise<void> {
  await setTimeOffset(page, days * 86400);
}

/**
 * Switch the deployment fallback language (debug only). Sets DEFAULT_LANGUAGE
 * (the per-user resolver falls back to it when a user has no pref) and clears
 * the iCal feed cache so the feed regenerates in the new language. `lang` is a
 * short code ('de','en',...); reset between tests with 'en'.
 */
export async function setLanguage(page: Page, lang: string): Promise<void> {
  const resp = await page.request.post('/debug/set_language', {
    data: { language: lang },
    headers: { 'Content-Type': 'application/json' },
  });
  if (!resp.ok()) {
    throw new Error(`set_language failed: HTTP ${resp.status()}`);
  }
}
